pkg/router/template/router.go

```diff
@@ -386,18 +386,27 @@ func (r *templateRouter) readState() error {
 // the state and refresh the backend. This is all done in the background
 // so that we can rate limit + coalesce multiple changes.
 func (r *templateRouter) Commit() {
-	r.lock.Lock()
-	defer r.lock.Unlock()
 
-	if !r.synced {
-		glog.V(4).Infof("Router state synchronized for the first time")
-		r.synced = true
-		r.stateChanged = true
-	}
+	needCommit := false
+
+	func() {
+		r.lock.Lock()
+		defer r.lock.Unlock()
+
+		if !r.synced {
+			glog.V(4).Infof("Router state synchronized for the first time")
+			r.synced = true
+			r.stateChanged = true
+		}
+
+		if r.stateChanged {
+			needCommit = true
+			r.stateChanged = false
+		}
+	}()
 
-	if r.stateChanged {
+	if needCommit {
```

Comment:
If I had time to review more I'd suggest simply refactoring rate limited commit function - it's not really well factored for the end goal.  But I don't have time for that.